### DISCUSSION TITLE
How to configure a datasource with custom fields

### DISCUSSION THREAD

**Discussion by maxdanilov** [renovatebot/renovate#16713] [Request Help] [answered]
Hello,

I'd like to implement an improvement for the Jenkins Plugins datasource. The idea is to enable the option to configure the current version of Jenkins a user is running, so a Renovate pull request is only created with the plugin updates that are compatible with this version. The current datasource behaviour is to create a pull request with the most recent versions of Jenkins plugins, regardless of whether they are compatible or not.

I stumbled upon a problem that I couldn't understand what is the proper way to pass a custom configuration item to configure a datasource. My naïve approach would be to have the ability to do something like that in the datasource configuration:

```json
  "packageRules": [
    {
      "datasources": ["jenkins-plugins"],
      "jenkinsVersion": "2.319.3"
    },
```

so `jenkinsVersion` is later available in the Jenkins Plugins datasource, to filter the available versions for a specific Jenkins version. if `jenkinsVersion` is omitted, the behaviour is the same as now.

However, as far as I understand, one cannot add a configuration field to an arbitrary datasource, and the only configuration possible is the items in the packageRules [options](https://docs.renovatebot.com/configuration-options/#packagerules) (for example, `registryUrls`)

More than that, by looking at the `Datasource` abstract class, every implementation of it (actual datasource classes) only supports `getReleases` with the following [signature](https://github.com/renovatebot/renovate/blob/68541455289c0ed93de316c9b5e1c1707f96c9e3/lib/modules/datasource/datasource.ts#L30-L32):   
```typescript
abstract getReleases(
    getReleasesConfig: GetReleasesConfig
  ): Promise<ReleaseResult | null>;
```

where `GetReleasesConfig` [is](https://github.com/renovatebot/renovate/blob/68541455289c0ed93de316c9b5e1c1707f96c9e3/lib/modules/datasource/types.ts#L21-L24):
```typescript
export interface GetReleasesConfig {
  packageName: string;
  registryUrl?: string;
}
```
meaning I cannot use any configuration parameters aside from `registryUrl`.

How should one be able to pass a configuration to a datasource if it's beyond the scope of the predefined possible configuration keys? Is it even possible?

**Comment by viceice** [comment, marked as answer]
it should be passed as extracted constraints / user configured constraints.

**Reply by viceice** [reply]
checkout pypi datasource, it will use the python constraints

**Reply by maxdanilov** [reply]
Thank you, @viceice. I've had a look into the `pypi` datasource, and I think I understand how to implement the change there. Would I need to do anything additional on the manager's side to enable the constraints to work? I couldn't understand by looking at the code of the existing managers like `poetry`

**Reply by viceice** [reply]
You need to fill `extractedConstraints` in manager `extractPackageFile`, that should automatically merged and passed to datasource as `constraints`

https://github.com/renovatebot/renovate/blob/56eb0bae9aa8d850f6d45981188c44de6e10fc23/lib/modules/manager/types.ts#L72

https://github.com/renovatebot/renovate/blob/56eb0bae9aa8d850f6d45981188c44de6e10fc23/lib/workers/repository/process/lookup/common.ts#L3-L14
